File: yabause/src/android/app/src/main/java/org/uoyabause/android/FileDialog.java
/*  Copyright 2019 devMiyax(smiyaxdev@gmail.com)

    This file is part of YabaSanshiro.

    YabaSanshiro is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    YabaSanshiro is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with YabaSanshiro; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
*/

package org.uoyabause.android;

import java.io.File;
import java.io.FilenameFilter;
import java.util.ArrayList;
import java.util.List;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Environment;
import android.util.Log;
import android.widget.Toast;

import java.util.logging.FileHandler;

import org.uoyabause.android.ListenerList.FireHandler;

class ListenerList<L> {
    private List<L> listenerList = new ArrayList<L>();

    public interface FireHandler<L> {
        void fireEvent(L listener);
    }

    public void add(L listener) {
        listenerList.add(listener);
    }

    public void fireEvent(FireHandler<L> fireHandler) {
        List<L> copy = new ArrayList<L>(listenerList);
        for (L l : copy) {
            fireHandler.fireEvent(l);
        }
    }

    public void remove(L listener) {
        listenerList.remove(listener);
    }

    public List<L> getListenerList() {
        return listenerList;
    }
}


public class FileDialog {
    private static final String PARENT_DIR = "..";
    private final String TAG = getClass().getName();
    private String[] fileList;
    private File currentPath;

    public interface FileSelectedListener {
        void fileSelected(File file);
    }

    public interface DirectorySelectedListener {
        void directorySelected(File directory);
    }

    private ListenerList<FileSelectedListener> fileListenerList = new ListenerList<FileDialog.FileSelectedListener>();
    private ListenerList<DirectorySelectedListener> dirListenerList = new ListenerList<FileDialog.DirectorySelectedListener>();
    private final Activity activity;
    private boolean selectDirectoryOption;
    private String fileEndsWith;

    /**
     * @param activity
     * @param path
     */
    public FileDialog(Activity activity, String path) {
        this.activity = activity; 
        loadFileList(new File(path));
    }

    /**
     * @return file dialog
     */
    public Dialog createFileDialog() {
        Dialog dialog = null;
        AlertDialog.Builder builder = new AlertDialog.Builder(activity);

        builder.setTitle(currentPath.getPath());

        if (selectDirectoryOption) {
         builder.setPositiveButton("Select directory", new DialogInterface.OnClickListener() {
         public void onClick(DialogInterface dialog, int which) { Log.d(TAG,
                 currentPath.getPath()); fireDirectorySelectedEvent(currentPath); }
         }); }


        builder.setItems(fileList, new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                String fileChosen = fileList[which];
                File chosenFile = getChosenFile(fileChosen);
                if (chosenFile.isDirectory()) {
                    loadFileList(chosenFile);
                    dialog.cancel();
                    dialog.dismiss();
                    showDialog();
                } else
                    fireFileSelectedEvent(chosenFile);
            }
        });
        builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) { 
                dialog.dismiss();
                fireFileSelectedEvent(null);
            }
        });
        dialog = builder.show();
        return dialog;
    }

    public void addFileListener(FileSelectedListener listener) {
        fileListenerList.add(listener);
    }

    public void removeFileListener(FileSelectedListener listener) {
        fileListenerList.remove(listener);
    }

    public void setSelectDirectoryOption(boolean selectDirectoryOption) {
        this.selectDirectoryOption = selectDirectoryOption;
    }

    public void addDirectoryListener(DirectorySelectedListener listener) {
        dirListenerList.add(listener);
    }

    public void removeDirectoryListener(DirectorySelectedListener listener) {
        dirListenerList.remove(listener);
    }

    /**
     * Show file dialog
     */
    public void showDialog() {
        createFileDialog().show();
    }

    private void fireFileSelectedEvent(final File file) {
        fileListenerList.fireEvent(new FireHandler<FileDialog.FileSelectedListener>() {
            public void fireEvent(FileSelectedListener listener) {
                listener.fileSelected(file);
            }
        });
    }

    private void fireDirectorySelectedEvent(final File directory) {
        dirListenerList.fireEvent(new FireHandler<FileDialog.DirectorySelectedListener>() {
            public void fireEvent(DirectorySelectedListener listener) {
                listener.directorySelected(directory);
            }
        });
    }

    private void loadFileList(File path) {
        if( path == null || path.isDirectory() == false ){
            path = new File(Environment.getExternalStorageDirectory(), "yabause");
            if (! path.exists()) path.mkdir();
        }
        this.currentPath = path;
        List<String> r = new ArrayList<String>();
        if (path.exists()) {
            if (path.getParentFile() != null)
                r.add(PARENT_DIR);

            FilenameFilter filter = new FilenameFilter() {
                public boolean accept(File dir, String filename) {
                    File sel = new File(dir, filename);
                    if (!sel.canRead())
                        return false;
                    if (selectDirectoryOption)
                        return sel.isDirectory();
                    else {
                        boolean endsWith = fileEndsWith != null ? filename
                                .toLowerCase().endsWith(fileEndsWith) : true;
                        return endsWith || sel.isDirectory();
                    }
                }
            };
            String[] fileList1 = path.list(filter);
            if(fileList1 != null){
                for (String file : fileList1) r.add(file);
            }else{
                Toast.makeText(this.activity,path.getAbsolutePath() + " is not readable. ",Toast.LENGTH_LONG );
            }
        }
        fileList = (String[]) r.toArray(new String[] {});
    }

    private File getChosenFile(String fileChosen) {
        if (fileChosen.equals(PARENT_DIR))
            return currentPath.getParentFile();
        else
            return new File(currentPath, fileChosen);
    }

    public void setFileEndsWith(String fileEndsWith) {
        this.fileEndsWith = fileEndsWith != null ? fileEndsWith.toLowerCase()
                : fileEndsWith;
    }
}

File: yabause/src/android/app/src/main/java/org/uoyabause/android/FileDialog.kt
/*  Copyright 2019 devMiyax(smiyaxdev@gmail.com)

    This file is part of YabaSanshiro.

    YabaSanshiro is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    YabaSanshiro is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with YabaSanshiro; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301  USA
*/
package org.uoyabause.android

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.util.Log
import android.widget.Toast
import java.io.File
import java.io.FilenameFilter
import java.util.ArrayList
import org.uoyabause.android.FileDialog.DirectorySelectedListener
import org.uoyabause.android.FileDialog.FileSelectedListener
import org.uoyabause.android.ListenerList.FireHandler
import org.uoyabause.android.YabauseStorage.Companion.storage
import java.util.Locale

internal class ListenerList<L> {
    private val listenerList: MutableList<L> = ArrayList()

    interface FireHandler<L> {
        fun fireEvent(listener: L)
    }

    fun add(listener: L) {
        listenerList.add(listener)
    }

    fun fireEvent(fireHandler: FireHandler<L>) {
        val copy: List<L> = ArrayList(listenerList)
        for (l in copy) {
            fireHandler.fireEvent(l)
        }
    }

    fun remove(listener: L) {
        listenerList.remove(listener)
    }

    fun getListenerList(): List<L> {
        return listenerList
    }
}

class FileDialog(private val activity: Activity, path: String?) {
    private val TAG = javaClass.name
    private var fileList: Array<String> = arrayOf<String>()
    private var currentPath: File? = null

    interface FileSelectedListener {
        fun fileSelected(file: File?)
    }

    interface DirectorySelectedListener {
        fun directorySelected(directory: File?)
    }

    private val fileListenerList = ListenerList<FileSelectedListener>()
    private val dirListenerList = ListenerList<DirectorySelectedListener>()
    private var selectDirectoryOption = false
    private var fileEndsWith: String? = null

    /**
     * @return file dialog
     */
    fun createFileDialog(): Dialog? {
        val builder = AlertDialog.Builder(activity)
        builder.setTitle(currentPath!!.path)
        if (selectDirectoryOption) {
            builder.setPositiveButton("Select directory") { _, _ ->
                Log.d(TAG,
                    currentPath!!.path)
                fireDirectorySelectedEvent(currentPath)
            }
        }
        builder.setItems(fileList) { dialog, which ->
            val fileChosen = fileList[which]
            val chosenFile = getChosenFile(fileChosen)
            if (chosenFile?.isDirectory == true) {
                loadFileList(chosenFile)
                dialog.cancel()
                dialog.dismiss()
                showDialog()
            } else fireFileSelectedEvent(chosenFile)
        }
        builder.setNegativeButton("Cancel") { dialog, _ ->
            dialog.dismiss()
            fireFileSelectedEvent(null)
        }
        return builder.show()
    }

    fun addFileListener(listener: FileSelectedListener) {
        fileListenerList.add(listener)
    }

    fun removeFileListener(listener: FileSelectedListener) {
        fileListenerList.remove(listener)
    }

    fun setSelectDirectoryOption(selectDirectoryOption: Boolean) {
        this.selectDirectoryOption = selectDirectoryOption
    }

    fun addDirectoryListener(listener: DirectorySelectedListener) {
        dirListenerList.add(listener)
    }

    fun removeDirectoryListener(listener: DirectorySelectedListener) {
        dirListenerList.remove(listener)
    }

    /**
     * Show file dialog
     */
    fun showDialog() {
        createFileDialog()!!.show()
    }

    private fun fireFileSelectedEvent(file: File?) {
        fileListenerList.fireEvent(object : FireHandler<FileSelectedListener> {
            override fun fireEvent(listener: FileSelectedListener) {
                listener.fileSelected(file)
            }
        })
    }

    private fun fireDirectorySelectedEvent(directory: File?) {
        dirListenerList.fireEvent(object : FireHandler<DirectorySelectedListener> {
            override fun fireEvent(listener: DirectorySelectedListener) {
                listener.directorySelected(directory)
            }
        })
    }

    private fun loadFileList(path: File?) {
        var lpath: File? = path
        if (lpath == null || lpath.isDirectory == false) {
            val s = storage
            lpath = File(s.rootPath)
            if (!lpath.exists()) lpath.mkdir()
        }
        currentPath = lpath
        val r: MutableList<String> = ArrayList()
        if (lpath.exists()) {
            if (lpath.parentFile != null) r.add(PARENT_DIR)
            val filter = FilenameFilter { dir, filename ->
                val sel = File(dir, filename)
                if (!sel.canRead()) return@FilenameFilter false
                if (selectDirectoryOption) sel.isDirectory else {
                    val endsWith = if (fileEndsWith != null) filename
                        .lowercase(Locale.getDefault()).endsWith(fileEndsWith!!) else true
                    endsWith || sel.isDirectory
                }
            }
            val fileList1 = lpath.list(filter)
            if (fileList1 != null) {
                for (file in fileList1) r.add(file)
            } else {
                Toast.makeText(activity,
                    lpath.absolutePath + " is not readable. ",
                    Toast.LENGTH_LONG)
            }
        }
        fileList = r.toTypedArray()
    }

    private fun getChosenFile(fileChosen: String): File? {
        return if (fileChosen == PARENT_DIR) currentPath!!.parentFile else File(
            currentPath,
            fileChosen)
    }

    fun setFileEndsWith(fileEndsWith: String?) {
        this.fileEndsWith = fileEndsWith?.lowercase(Locale.getDefault()) ?: fileEndsWith
    }

    companion object {
        private const val PARENT_DIR = ".."
    }

    /**
     * @param activity
     * @param path
     */
    init {
        loadFileList(File(path!!))
    }
}
